File: plugins/toonkit/skills/3dref/scripts/timewarp.py
"""C2 speed ramps integrated exactly. Output seconds -> shared action seconds."""
import math

class TimeWarp:
    def __init__(self,rows,duration):
        self.rows=rows if rows is not None else [{'time':0,'speed':1},{'time':duration,'speed':1}]
        if len(self.rows)<2: raise ValueError('timeWarp needs two or more speed anchors')
        for r in self.rows:
            if set(r)!={'time','speed'} or any(type(r[k]) not in (int,float) or not math.isfinite(r[k]) for k in r): raise ValueError('Invalid timeWarp anchor')
            if r['speed']<=0: raise ValueError('timeWarp speed must be positive; choreograph explicit holds separately')
        if self.rows[0]['time']!=0 or abs(self.rows[-1]['time']-duration)>1e-8: raise ValueError('timeWarp must span 0..durationSeconds')
        if any(b['time']<=a['time'] for a,b in zip(self.rows,self.rows[1:])): raise ValueError('timeWarp times must increase')
        self.duration=duration

    def at(self,t):
        if not 0<=t<=self.duration+1e-9: raise ValueError('timeWarp sample out of range')
        total=0.
        for a,b in zip(self.rows,self.rows[1:]):
            dt=b['time']-a['time'];u=min(1,max(0,(t-a['time'])/dt))
            # integral of 6u^5 - 15u^4 + 10u^3 = u^6 - 3u^5 + 2.5u^4
            total+=dt*(a['speed']*u+(b['speed']-a['speed'])*(u**6-3*u**5+2.5*u**4))
            if t<=b['time']: return total,a['speed']+(b['speed']-a['speed'])*u**3*(10+u*(-15+6*u))
        return total,self.rows[-1]['speed']

    @property
    def identity(self): return all(r['speed']==1 for r in self.rows)

File: plugins/toonkit/skills/3dref/scripts/timewarp.mjs
/** Paired with timewarp.py; parity tested at ramp boundaries and every frame. */
export function timeWarp(rows,duration) {
  rows=rows??[{time:0,speed:1},{time:duration,speed:1}];
  if(!Array.isArray(rows)||rows.length<2||rows.some((r,i)=>Object.keys(r).sort().join(',')!=='speed,time'||!Number.isFinite(r.time)||!Number.isFinite(r.speed)||r.speed<=0||(i&&r.time<=rows[i-1].time))||rows[0].time!==0||Math.abs(rows.at(-1).time-duration)>1e-8)throw Error('Invalid timeWarp: positive speed anchors must span 0..durationSeconds');
  const at=t=>{
    if(!Number.isFinite(t)||t<0||t>duration+1e-9)throw Error('timeWarp sample out of range');
    let total=0;
    for(let i=0;i<rows.length-1;i++){
      const a=rows[i],b=rows[i+1],dt=b.time-a.time,u=Math.min(1,Math.max(0,(t-a.time)/dt));
      total+=dt*(a.speed*u+(b.speed-a.speed)*(u**6-3*u**5+2.5*u**4));
      if(t<=b.time)return {time:total,speed:a.speed+(b.speed-a.speed)*u**3*(10+u*(-15+6*u))};
    }
    return {time:total,speed:rows.at(-1).speed};
  };
  return {at,identity:rows.every(r=>r.speed===1),rows};
}
